### ISSUE TITLE
Table span failures due to empty sequence in variables

### ISSUE THREAD

**Issue by takesson** [dita4publishers/org.dita4publishers.dita2indesign#8]
We have noticed a few situations when table spans can make the transform to Icml fail with message like:
```
calstbl2IcmlImpl.xsl:166: Fatal Error! An empty sequence is not allowed as the value of variable $colNumber
```

We have created 4 example topics with minimal tables that reproduce the issues. I have committed them in a fork of this repository.

The naming of the test files are based on table size and span size (COLxROW size):
 - table-1x2-span-1x2-test.dita fails in variable $baseWid (Line#: 378). There is a superfluous (?) entry in row 2 that some editors leave behind when spanning. Fails also when removing that entry.
 - table-2x1-span-2x1-test.dita fails in variable $baseWid.
 - table-2x2-span-2x2-test.dita fails in variable $baseWid.
 - table-3x2-span-2x2-test.dita fails in variable $colNumber (Line#: 166). This is the issue we have seen on actual authored data.

I have looked at the code for the table matrix but struggling to fully understand the implementation. Can we somehow work together to resolve these failures?

**Comment by takesson** [comment]
I can create a PR with the test files if you find them valid for inclusion.

**Comment by drmacro** [comment]
I am on holiday this week with limited internet access. I can take a look at this at some point. But first ensure that your tables are valid in terms of consistent column definition and span definitions. Some processors are more or less tolerant so may be accepting invalid tables. The latest Oxygen has improved CALS table validation.

**Comment by drmacro** [comment]
Can you either add me to your fork or create a gist with the failing tables?

**Comment by takesson** [comment]
I am sorry, I haven't had time to validate the tables in Oxygen. I have made them with Arbortext and validated them there. I was unable to see anything obvious by inspecting them manually.

I have tried to narrow down the cases to as small tables as possible to simplify debugging.

Thanks for looking at the issue!

**Comment by drmacro** [comment]
I have replicated the $baseWid failure. Investigating.

**Comment by drmacro** [comment]
One issue: these tables don't have column width specifications. Without column widths there's no obvious way to generate useful InDesign tables. 

That's why the $baseWid failure is happening: it expects there to be a @colwidth attribute on the colspec element.

**Comment by takesson** [comment]
Ok, understood.

When adding a colwidth to table-1x2-span-1x2-test.dita it instead produces the `$colNumber` error. Do you see the same result?

I thought the lack of colwidth attributes would basically mean "evenly sized columns". Would it make sense to use a default value, e.g. `1.0*` when colwidth is empty?

**Comment by drmacro** [comment]
Yes, best to specify "*" or "1*" if you want all the columns to be the same width.

**Comment by drmacro** [comment]
Reviewed the CALS exchange table docs. @colwidth is option and the default is '1*' so updating code to handle that case.

**Comment by drmacro** [comment]
OK, I've fixed the failure and pushed updates to master and develop branch.

The failure was caused by having rows spanned into that didn't need to actually be there, that is, a row that both spans all cells horizontally and spans the following row. The cell in the following row does not contribute to the rendered table and so wasn't reflected in the intermediate matrix table.

I corrected the code by simply ignoring the cell in that case.

For omitted @colwidth I added a check and set the width to "1*" if it's not specified.

**Comment by takesson** [comment]
Thanks Eliot! The transforms complete without failure now.

>  The cell in the following row does not contribute to the rendered table and so wasn't reflected in the intermediate matrix table.

I think your analysis is correct for 3 of the sample document but table-3x2-span-2x2-test.dita does not have any entry elements that should not be rendered. I have "Placed" that Indesign snippet now and the result is that the cells "value1" and "value2" are lost/not displayed.

I suspect there is some additional complexity.
